src/common/random_generator.h

```diff
@@ -0,0 +1,224 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * Copyright (c) 2017 by Contributors
+ * \file random_generator.h
+ * \brief Native random number generator.
+ */
+#ifndef MXNET_COMMON_RANDOM_GENERATOR_H_
+#define MXNET_COMMON_RANDOM_GENERATOR_H_
+
+#include <mxnet/base.h>
+#include <random>
+#include <new>
+
+#if MXNET_USE_CUDA
+#include <cuda.h>
+#include <cuda_runtime.h>
+#include <curand_kernel.h>
+#endif  // MXNET_USE_CUDA
+
+using namespace mshadow;
+
+namespace mxnet {
+namespace common {
+namespace random {
+
+template<typename Device, typename DType MSHADOW_DEFAULT_DTYPE>
+class RandGenerator;
+
+// Elementary random number generation for int/uniform/gaussian in CPU and GPU.
+// Will use float data type whenever instantiated for half_t or any other non
+// standard real type.
+template<typename Device, typename DType MSHADOW_DEFAULT_DTYPE>
+class RandGeneratorImpl;
+
+// at least how many random numbers should be generated by one CPU thread.
+const int kCPUMinRndNumberPerThread = 64;
+// store how many global random states for CPU.
+const int kCPURndStateNum = 1024;
+
+template<typename DType>
+class RandGeneratorImpl<cpu, DType> {
+ public:
+  typedef typename std::conditional<std::is_floating_point<DType>::value,
+                                    DType, double>::type FType;
+
+  explicit RandGeneratorImpl<cpu, DType>(std::mt19937 *ptr_engine) : engine_(ptr_engine) {}
+
+  MSHADOW_XINLINE int rand() { return engine_->operator()(); }
+
+  MSHADOW_XINLINE FType uniform() {
+    typedef typename std::conditional<std::is_integral<DType>::value,
+                                      std::uniform_int_distribution<DType>,
+                                      std::uniform_real_distribution<FType>>::type GType;
+    GType dist_uniform;
+    return dist_uniform(*engine_);
+  }
+
+  MSHADOW_XINLINE FType normal() {
+    std::normal_distribution<FType> dist_normal;
+    return dist_normal(*engine_);
+  }
+
+ private:
+  std::mt19937 *engine_;
+};
+
+template<typename DType>
+class RandGenerator<cpu, DType> {
+ public:
+  MSHADOW_XINLINE void dispose() {}
+
+  MSHADOW_XINLINE RandGeneratorImpl<cpu, DType> Get(int idx = 0) {
+    std::mt19937 *ptr_engine = &states_[idx];
+    RandGeneratorImpl<cpu, DType> gen(ptr_engine);
+    return gen;
+  }
+
+  MSHADOW_XINLINE void Seed(Stream<cpu> *, uint32_t seed) {
+    for (int i = 0; i < kCPURndStateNum; ++i) (states_ + i)->seed(seed + i);
+  }
+
+  MSHADOW_XINLINE void set_state(int idx, const std::mt19937 &state) {
+    states_[idx] = state;
+  }
+
+ private:
+  std::mt19937 states_[kCPURndStateNum];
+};
+
+#if MXNET_USE_CUDA
+
+// at least how many random numbers should be generated by one GPU thread.
```

Comment:
why do we need this?

Reply:
a number of contiguous random numbers should be generated from one state